File: src/main/java/com/NGSViz/cli/CliResponse.java
package com.NGSViz.cli;

import com.NGSViz.compute.ComputeValidationResult;
import com.NGSViz.compute.ValidationIssue;
import org.json.JSONArray;
import org.json.JSONObject;

import java.util.List;

/** Machine-readable response envelope for native CLI commands. */
public final class CliResponse {
    private static final String SCHEMA_VERSION = "1.0";

    private final String command;
    private final String status;
    private final int exitCode;
    private final String message;
    private final Object data;
    private final JSONArray warnings;
    private final JSONArray errors;

    private CliResponse(String command, String status, int exitCode, String message,
                        Object data, JSONArray warnings, JSONArray errors) {
        this.command = command;
        this.status = status;
        this.exitCode = exitCode;
        this.message = message;
        this.data = data;
        this.warnings = warnings;
        this.errors = errors;
    }

    public static CliResponse success(String command, String message, JSONObject data) {
        return new CliResponse(
                command, "success", CliExitCode.SUCCESS.code(), message, data,
                new JSONArray(), new JSONArray()
        );
    }

    public static CliResponse error(String command, String message, String errorCode) {
        return issueResponse(command, "error", message, errorCode);
    }

    public static CliResponse unsupported(String command) {
        return issueResponse(
                command, "unsupported",
                "Command is not implemented in Phase 3.",
                "COMMAND_UNSUPPORTED"
        );
    }

    public static CliResponse requestError(String command, String message,
                                           JSONObject data, List<ValidationIssue> issues) {
        return new CliResponse(
                command, "error", CliExitCode.USAGE_ERROR.code(), message, data,
                new JSONArray(), issuesJson(issues)
        );
    }

    public static CliResponse validation(String command, JSONObject data,
                                         ComputeValidationResult result) {
        String status = result.isValid() ? "success" : "error";
        String message = result.isValid() ? "Compute request is valid."
                : "Compute request validation failed.";
        return new CliResponse(
                command, status, result.exitCode(), message, data,
                result.warningsJson(), result.errorsJson()
        );
    }

    static CliResponse computation(String status, int exitCode, String message,
                                   JSONObject data, List<ValidationIssue> warnings,
                                   List<ValidationIssue> errors) {
        return new CliResponse(
                CliCommand.RUN_COMPUTE.commandName(), status, exitCode, message, data,
                issuesJson(warnings), issuesJson(errors)
        );
    }

    private static JSONArray issuesJson(List<ValidationIssue> issues) {
        JSONArray json = new JSONArray();
        for (ValidationIssue issue : issues) json.put(issue.toJson());
        return json;
    }

    private static CliResponse issueResponse(String command, String status,
                                             String message, String errorCode) {
        JSONObject error = new JSONObject();
        error.put("code", errorCode);
        error.put("field", JSONObject.NULL);
        error.put("message", message);
        return new CliResponse(
                command, status, CliExitCode.USAGE_ERROR.code(), message,
                JSONObject.NULL, new JSONArray(), new JSONArray().put(error)
        );
    }

    public int exitCode() {
        return exitCode;
    }

    public JSONObject toJson() {
        JSONObject response = new JSONObject();
        response.put("schema_version", SCHEMA_VERSION);
        response.put("command", command);
        response.put("status", status);
        response.put("exit_code", exitCode);
        response.put("message", message);
        response.put("data", data);
        response.put("warnings", warnings);
        response.put("errors", errors);
        return response;
    }
}

File: lib/cli/cli_response.R
new_cli_issue <- function(code, file = NULL, field = NULL, message) {
  list(code = code, file = file, field = field, message = message)
}

build_cli_response <- function(command, status, exit_code, message, data = NULL,
                               warnings = list(), errors = list()) {
  list(
    schema_version = "1.0",
    command = command,
    status = status,
    exit_code = as.integer(exit_code),
    message = message,
    data = data,
    warnings = unname(warnings),
    errors = unname(errors)
  )
}

build_validation_response <- function(command, input_dir, environment_result,
                                      validation_result = NULL) {
  input_path <- normalizePath(input_dir, mustWork = FALSE)
  if (!environment_result$valid) {
    data <- list(
      input_dir = input_path,
      r_version = environment_result$r_version,
      missing_packages = unname(environment_result$missing_packages)
    )
    return(build_cli_response(
      command, "error", 3L, "Visualization environment validation failed.",
      data, errors = environment_result$errors
    ))
  }
  data <- list(
    input_dir = input_path,
    plot_setting_count = validation_result$plot_setting_count,
    manifest = NULL,
    r_version = environment_result$r_version
  )
  if (!validation_result$valid) {
    return(build_cli_response(
      command, "error", validation_result$exit_code,
      "Visualization input validation failed.", data,
      validation_result$warnings, validation_result$errors
    ))
  }
  build_cli_response(
    command, "success", 0L, "Visualization inputs are valid.", data,
    validation_result$warnings
  )
}

emit_cli_response <- function(response) {
  cat(jsonlite::toJSON(
    response, auto_unbox = TRUE, null = "null", na = "null", digits = NA
  ))
  cat("\n")
}
